### PR TITLE
Kiera - RideShare - Octos

### PR DESCRIPTION
**Pull request by Krashaune**
# Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | I am not quite sure if this fits within design but I needed to calculate time an use it more than once. At first I wrote the code in driver and passenger where it was needed but, that was duplicating so, I added a self method in trips to do that work and I used it in driver and passenger which helped with drying up my code.
Describe any examples of composition that you encountered in this project, if any | If I understand that composition is about the structure of the program. I encountered it a lot as I was trying to figure out how all the classes either worked or got information from each other. 
Describe a _nominal test_ that you wrote for this assignment. | A nominal test I wrote was for the trip dispatcher request trip method it raised and argument error if the passenger id was not found.
Describe an _edge case test_ that you wrote for this assignment. | I don't think i wrote any edge cases. But if I did I would have wrote one in trip dispatcher for if there are no trips in than the id for request trip would start at one. Now it is looking at the last id and adding one to it to create the next trip id.
Describe a concept that you gained more clarity on as you worked on this assignment. | I gained a little more clarity on how to use methods from other classes as well as how to effectively write a helper method to do a small thing that doesn't quite make sense in another method. I also gained more clarity on why it is important to understand what the output of everything is before trying to access the data.

### CODE REVIEW

File: lib/driver.rb [hunk 1]
```diff
@@ -42,6 +43,34 @@ def add_trip(trip)
...
+
+    # calulate drivers average revenue per hour spend driving
+    def average_revenue
+      # input: total_revenue, total_time_spent(hours),
+      # output: average_revenue = total_revenue / total_time_spent
+      average_revenue = 0
+      total_time = Trip.total_time(trips)/60/60.round(2)
+      average = self.total_revenue / total_time
+      average_revenue += average
+      return average_revenue.round(2)
```
Comment (droberts-sea) [inline lib/driver.rb:68]: If you're rounding the final value (line 68), you shouldn't need to also round an intermediate value (line 65).

Also, I believe the `.` will bind tighter than the `/`, so the `round(2)` will be applied to the last `60`. If you wanted to round the whole thing you would have to use parentheses: `(Trip.total_time(trips)/60/60).round(2)`

File: lib/driver.rb [hunk 2]
```diff
@@ -42,6 +43,34 @@ def add_trip(trip)
...
+    end
+
+    def change_status
+      @status = :UNAVAILABLE
+      return
+    end # change_status
```
Comment (droberts-sea) [inline lib/driver.rb:74]: A more descriptive name for this method might be `set_unavailable`.

File: lib/driver.rb [hunk 3]
```diff
@@ -42,6 +43,34 @@ def add_trip(trip)
...
+
+    def total_revenue
+      tax = 1.65
+      driver_pay = 0.8
+      gross_pay = 0
+
+      trips.each do |trip|
+        gross_pay += trip.cost - tax
+      end
+      total_revenue = gross_pay * driver_pay
```
Comment (droberts-sea) [inline lib/driver.rb:56]: This method will hit an error if you try to run it on a driver with an incomplete trip. There are a couple of workarounds for this.

You could explicitly ignore trips with a `nil` cost:

```ruby
trips.each do |trip|
  if trip.cost.nil?
    next
  end
  # ... add to the total ...
end
```

Or even better, you could write a helper method that returns a list of complete trips:

```ruby
def completed_trips
  return @trips.reject{ |t| t.end_time.nil? }
end

def total_revenue
  completed_trips.each do |trip|
    # ... same logic as before ...
  end
end
```

File: lib/trip.rb [hunk 4]
```diff
@@ -12,10 +12,35 @@ def initialize(input)
...
+    def duration
+    # calculate the duration of the trip in seconds
+      duration = (@end_time - @start_time)
+    return duration
+    end
+
+    def self.total_time(trip_list)
+      total_time = 0
+      trip_list.each do |trip|
+       total_time += trip.duration
```
Comment (droberts-sea) [inline lib/trip.rb:36]: Good work DRYing up this code!

File: specs/driver_spec.rb [hunk 5]
```diff
@@ -76,5 +81,53 @@
...
-end
+  end # average_rating
+
+
+  describe 'total_revenue' do
+    it "calculates take home pay for driver" do
+      trips = [
+        RideShare::Trip.new({cost: 10, rating: 3, start_time: Time.parse('2015-05-20T14:14:00+00:00'),
+          end_time: Time.parse('2015-05-20T14:30:00+00:00')}),
+        RideShare::Trip.new({cost: 4, rating: 3, start_time: Time.parse('2015-05-20T14:14:00+00:00'),
```
Comment (droberts-sea) [inline specs/driver_spec.rb:92]: This is just the sort of thing you could DRY up using a `before` block. If you put it in the outermost `describe` (the one for `Driver`), it will be available inside all the nested `describe`s (for the individual methods).

File: specs/driver_spec.rb [hunk 6]
```diff
@@ -76,5 +81,53 @@
...
+      }
+
+      driver = RideShare::Driver.new(data)
+      driver.total_revenue.must_equal 12.04
+    end # calculate total rev
+  end # total_revenue
+
+  describe 'average_revenue' do
+    it "can calculate the average pay per hour spent driving" do
+
```
Comment (droberts-sea) [inline specs/driver_spec.rb:112]: Other questions that I would like to see these tests ask, for all the wave 1 methods (on both `Driver` and `Passenger`):

- What happens if the driver or passenger has no trips?
- What happens if the driver or passenger has an incomplete trip?

File: specs/trip_dispatch_spec.rb [hunk 7]
```diff
@@ -88,5 +89,40 @@
...
+      new_request.must_be_instance_of RideShare::Trip
+      new_request.driver.status.must_equal :UNAVAILABLE
+      new_request.driver.trips.must_include new_request
+      new_request.passenger.trips.must_include new_request
+
+    end # driver AVAILABLE
+
+    it "raises an ArgumentError if passenger_id is not found" do
+      dispatcher = RideShare::TripDispatcher.new
+      passenger_id = dispatcher.passengers.last.id + 1
```
Comment (droberts-sea) [inline specs/trip_dispatch_spec.rb:118]: There are many more things you should be testing for the `request_trip` method:

- Does the trip reference the correct passenger and driver?
- Was the trip added to the trip lists of the driver and passenger?
- What if there is no available driver?

The last is an important edge case, and is exactly the sort of situation where software could do something unexpected. You might set up the test as follows:

```ruby
it 'raises an error if no driver is available' do
  dispatcher = RideShare::TripDispatcher.new
  dispatcher.drivers.each do |driver|
    # This may require an attr_writer in Driver
    driver.status = :UNAVAILABLE
  end
  proc {
    dispatcher.request_trip(2)
  }.must_raise StandardError
end
```

**Comment by droberts-sea** [comment]
# Ride Share
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | yes
Answer comprehension questions | yes - Your answer about composition gets at this, but I want to call out this vocab explicitly. By composition we mean a _has-a_ or _has-many_ relationship. Examples in this project include "a TripDispatcher _has-many_ Drivers, Passengers and Trips", "a Driver _has-many_ Trips", "a Trip _has-a_ Driver".
**Wave 1** |
`Trip` has a helper method to calculate duration | yes
`Passenger` has a method to calculate total cost of all trips | yes
`Driver` has a method to calculate total revenue | yes
Complete tests | some missing - see inline
**Wave 2** |
`TripDispatcher` has a new method to create trips | yes
Creating a trip in `TripDispatcher` relies on methods in `Driver` and `Passenger` to modify their own attributes | yes
Methods from wave 1 work on `Driver`s and `Passenger`s with incomplete trips | no - see inline
Complete tests | some missing - see inline
**Wave 3** (optional) |
Complex logic was correctly implemented and tested | N/A
Appropriate use of Ruby's `Date` | N/A
Appropriate helper methods were made to help with complex logic | N/A
Complete tests | N/A
**Overall** | 

This is a good start. Your code is clean and readable, and I really like the way you've broken your code up with helper methods.

There big area where I see room for improvement is test coverage. You're doing a good job of identifying nominal cases, but as you mentioned in the comprehension questions I'm not seeing any edge cases at all. You should work on finding the "edge" between working and not working, which means identifying the spots where a method you've written might do something funny. I've tried to outline some of these in the comments below.

That being said, your product code looks quite solid, and you've certainly solved the problem at hand. I'm excited to see what you come up with on Hotel. Keep up the hard work!
